src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/ReadWriteSplittingPlugin.java

```diff
@@ -147,22 +242,71 @@ private void updateTopology(JdbcConnection currentConnection, HostInfo currentHo
     }
 
     final int oldHostIndex = this.currentHostIndex;
-    final int newHostIndex = this.rdsHostUtils.getHostIndex(this.hosts, currentHost);
+    final int newHostIndex = getHostIndex(
+            currentConnection,
+            currentHost,
+            Messages.getString("ReadWriteSplittingPlugin.5"));
+
     if (oldHostIndex == WRITER_INDEX && newHostIndex > WRITER_INDEX) {
       // The old host was a writer; we have failed over to a new writer.
       // Update the host list with the new writer at the beginning.
       Collections.swap(this.hosts, oldHostIndex, newHostIndex);
     }
   }
 
+  void pickNewReaderConnection() {
+    if (this.hosts.size() <= 2) {
+      return;
+    }
+
+    final ArrayDeque<Integer> readerIndexes = getRandomReaderIndexes();
+    while (!readerIndexes.isEmpty()) {
+      final int index = readerIndexes.poll();
+
+      try {
+        this.readerConnection = getConnectionToHost(index);
+        final JdbcConnection currentConnection = this.currentConnectionProvider.getCurrentConnection();
+        switchCurrentConnectionTo(currentConnection, this.readerConnection, index);
+        return;
+      } catch (SQLException e) {
```

Comment:
Does it makes sense to continue with while-loop in case exception is communication one, and rethrow it otherwise?